### PR TITLE
Add a method to CompositeChange to return a flattened version

### PR DESCRIPTION
**Pull request by trevan**

### CODE REVIEW

File: game-core/src/test/java/games/strategy/engine/data/CompositeChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import static org.mockito.Mockito.when;
+
+import games.strategy.engine.data.changefactory.ChangeFactory;
+import java.util.List;
+import org.junit.jupiter.api.Test;
+
+class CompositeChangeTest {
+
+  @Test
+  void flattenChanges() {
```
Comment (DanVanAtta) [inline game-core/src/test/java/games/strategy/engine/data/CompositeChangeTest.java:15]: Couple more test cases to consider having:
- trivial case where there is a single composite change and the input is the same as output
- deeply nested (more than 2) singleton change

File: game-core/src/main/java/games/strategy/engine/data/CompositeChange.java [hunk 2]
```diff
@@ -22,6 +24,21 @@ public CompositeChange(final List<Change> changes) {
     this.changes = new ArrayList<>(changes);
   }
 
+  public CompositeChange flatten() {
+    return new CompositeChange(
+        changes.stream()
+            .map(
+                change -> {
+                  if (change instanceof CompositeChange) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/engine/data/CompositeChange.java:32]: nit: A ternary would be worth consider having here, would make the lambda syntax a bit nicer.

File: game-core/src/main/java/games/strategy/engine/data/CompositeChange.java [hunk 3]
```diff
@@ -22,6 +24,21 @@ public CompositeChange(final List<Change> changes) {
     this.changes = new ArrayList<>(changes);
   }
 
+  public CompositeChange flatten() {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/engine/data/CompositeChange.java:27]: Could use a javadoc IMO, particularly considering the method is recursive.

**Comment by trevan** [comment]
@DanVanAtta , I've added some documentation and added another test case.

File: game-core/src/test/java/games/strategy/engine/data/CompositeChangeTest.java (hunk 1 above)
Comment (trevan) [reply game-core/src/test/java/games/strategy/engine/data/CompositeChangeTest.java:15]: The existing test case already is testing deeply nested changes.  It goes at least 3 deep.

Comment (DanVanAtta) [reply game-core/src/test/java/games/strategy/engine/data/CompositeChangeTest.java:15]: Indeed, I'm drawing a slight distinction:
- deeply nested with N+1 changes
- deeply nested with just 1 change

In the latter case one would expect that to be flattened into a single change.

Comment (trevan) [reply game-core/src/test/java/games/strategy/engine/data/CompositeChangeTest.java:16]: @DanVanAtta , I see what you mean.  I've added a test for the later case.

**Comment by DanVanAtta** [comment]
:+1:

**Pull request merged**